xarray/backends/common.py

```diff
@@ -216,7 +216,11 @@ def store_dataset(self, dataset):
 
     def store(self, variables, attributes, check_encoding_set=frozenset(),
               unlimited_dims=None):
+        # This seems out of place
+        variables = OrderedDict([(k, maybe_encode_as_char_array(v))
+                                 for k, v in variables.items()])
```

Comment:
Should each backend define encode/decode methods? For string variables, we need to encode before setting the dimensions.